framework/src/meshmodifiers/ElementSubdomainModifierBase.C

```diff
@@ -0,0 +1,632 @@
+  _reinitialized_nodes.clear();
+
+  for (const auto & [elem_id, subdomain] : moved_elems)
+  {
+    // Default: any element that changes subdomain is reinitialized
+    if (std::find(_subdomain_ids_to_reinitialize.begin(),
+                  _subdomain_ids_to_reinitialize.end(),
+                  Moose::ANY_BLOCK_ID) != _subdomain_ids_to_reinitialize.end())
+      _reinitialized_elems.insert(elem_id);
+    else // Reinitialize if new subdomain is in list of subdomains to be reinitialized
+    {
+      const auto & [from, to] = subdomain;
+      if (subdomainIsReinitialized(to) && _previous_subdomain_reinitialized)
+        _reinitialized_elems.insert(elem_id);
+      // Only reinitialize if original subdomain is not in list of subdomains
+      else if (subdomainIsReinitialized(to) && !_previous_subdomain_reinitialized &&
+               !subdomainIsReinitialized(from))
+        _reinitialized_elems.insert(elem_id);
+      else // New subdomain is not in list of subdomains
+        continue;
+    }
+
+    const auto elem = _mesh.elemPtr(elem_id);
+    for (unsigned int i = 0; i < elem->n_nodes(); ++i)
+      if (nodeIsNewlyReinitialized(elem->node_id(i)))
+        _reinitialized_nodes.insert(elem->node_id(i));
+  }
+}
+
+bool
+ElementSubdomainModifierBase::subdomainIsReinitialized(SubdomainID id) const
+{
+  // Default: any element that changes subdomain is reinitialized
+  if (std::find(_subdomain_ids_to_reinitialize.begin(),
+                _subdomain_ids_to_reinitialize.end(),
+                Moose::ANY_BLOCK_ID) != _subdomain_ids_to_reinitialize.end())
+    return true;
+
+  // Is subdomain in list of subdomains to be reinitialized
+  return std::find(_subdomain_ids_to_reinitialize.begin(),
+                   _subdomain_ids_to_reinitialize.end(),
+                   id) != _subdomain_ids_to_reinitialize.end();
+}
+
+bool
+ElementSubdomainModifierBase::nodeIsNewlyReinitialized(dof_id_type node_id) const
+{
+  // If any of the node neighbors are already reinitialized, then the node is NOT newly
+  // reinitialized.
+  for (auto neighbor_elem_id : _mesh.nodeToElemMap().at(node_id))
+    if (subdomainIsReinitialized(_mesh.elemPtr(neighbor_elem_id)->subdomain_id()))
+      return false;
+  return true;
+}
+
+void
+ElementSubdomainModifierBase::applyIC(bool displaced)
+{
+  _fe_problem.projectInitialConditionOnCustomRange(reinitializedElemRange(displaced),
+                                                   reinitializedBndNodeRange(displaced));
+
+  // Set old and older solutions on the reinitialized dofs to the reinitialized values
+  setOldAndOlderSolutions(_fe_problem.getNonlinearSystemBase(_sys.number()),
+                          reinitializedElemRange(displaced),
+                          reinitializedBndNodeRange(displaced));
+  setOldAndOlderSolutions(_fe_problem.getAuxiliarySystem(),
+                          reinitializedElemRange(displaced),
+                          reinitializedBndNodeRange(displaced));
```

Comment:
Will it be helpful to add an option to not update old and older solutions here?

Reply:
Currently, if an old solution does not exist, then the `setOldAndOlderSolutions` isn't carried out. 

If the user specifies that reinitialization occurs in an element that changes subdomain, it was assumed that the old solution of the previous subdomain wasn't required anymore, as the element would be a 'new' material, e.g. solid to liquid. 

There is still the option to turn off reinitialization completely, if the user only wants to change subdomain but wants to keep all the values the same. 

Is there a scenario you're thinking of where it would be useful to reinitialize an element value, but still keep the old and older solutions as the values from the previous subdomain? If there is, then that option could be added in.

Reply:
I was thinking about phase transition too. But you are right, it makes more sense to reinit both old and older solutions here. I don't have a strong application case in mind at the moment, so it might be best to leave it as is!